java-checks/src/main/java/org/sonar/java/checks/PatternMatchUsingIfCheck.java

```diff
@@ -0,0 +1,280 @@
+      expr = ((BinaryExpressionTree) expr).leftOperand();
+    }
+    return expr;
+  }
+
+  private static void populateGuardsList(ExpressionTree expr, Deque<ExpressionTree> guards) {
+    while (expr instanceof BinaryExpressionTree binary && binary.kind() == Tree.Kind.CONDITIONAL_AND) {
+      guards.addFirst(binary.rightOperand());
+      expr = binary.leftOperand();
+    }
+  }
+
+  private static boolean isElseIf(IfStatementTree ifStat) {
+    return ifStat.parent() instanceof IfStatementTree parentIf && parentIf.elseStatement() == ifStat;
+  }
+
+  private JavaQuickFix computeQuickFix(List<Case> cases, IfStatementTree topLevelIfStat) {
+    var baseIndent = topLevelIfStat.firstToken().range().start().column() - 1;
+    var sb = new StringBuilder();
+    sb.append("switch (").append(cases.get(0).scrutinee()).append(") {\n");
+    for (Case caze : cases) {
+      sb.append(" ".repeat(baseIndent + INDENT));
+      writeCase(caze, sb, baseIndent);
+      sb.append("\n");
+    }
+    sb.append(" ".repeat(baseIndent)).append("}");
+    var edit = JavaTextEdit.replaceTree(topLevelIfStat, sb.toString());
+    return JavaQuickFix.newQuickFix(ISSUE_MESSAGE).addTextEdit(edit).build();
+  }
+
+  private void writeCase(Case caze, StringBuilder sb, int baseIndent) {
+    if (caze instanceof PatternMatchCase patternMatchCase) {
+      sb.append("case ").append(QuickFixHelper.contentForTree(patternMatchCase.pattern, context));
+      if (!patternMatchCase.guards().isEmpty()) {
+        List<ExpressionTree> guards = patternMatchCase.guards();
+        sb.append(" when ");
+        join(guards, " && ", sb);
+      }
+    } else if (caze instanceof EqualityCase equalityCase) {
+      sb.append("case ");
+      join(equalityCase.constants, ", ", sb);
+    } else {
+      sb.append("default");
+    }
+    sb.append(" -> ");
+    addIndentedExceptFirstLine(makeBlockCode(caze.body(), baseIndent), sb);
+  }
+
+  private String makeBlockCode(StatementTree stat, int baseIndent){
+    var rawCode = QuickFixHelper.contentForTree(stat, context);
+    if (stat instanceof BlockTree){
+      return rawCode;
+    } else {
+      return "{\n" + " ".repeat(baseIndent + INDENT) + rawCode + "\n" + " ".repeat(baseIndent) + "}";
+    }
+  }
+
+  private static void addIndentedExceptFirstLine(String s, StringBuilder sb) {
+    var lines = s.lines().iterator();
+    if (!lines.hasNext()) {
+      return;
+    }
+    var indentStr = " ".repeat(INDENT);
+    sb.append(lines.next()).append("\n");
+    while (lines.hasNext()) {
+      sb.append(indentStr).append(lines.next());
+      if (lines.hasNext()) {
+        sb.append("\n");
+      }
+    }
+  }
+
+  private void join(List<? extends Tree> elems, String sep, StringBuilder sb) {
+    var iter = elems.iterator();
+    while (iter.hasNext()) {
+      var e = iter.next();
+      sb.append(QuickFixHelper.contentForTree(e, context));
+      if (iter.hasNext()) {
+        sb.append(sep);
+      }
+    }
```

Comment:
My reasoning was that if `elems` is a `LinkedList` (which is the case with the current implementation), the solution using `get` makes the complexity non-linear

Reply:
Do we need the LinkedList inside `extractCasesFromIfSequence` ? It should work the same with an ArrayList right?

Reply:
I used a LinkedList because we are always adding the element at the end, and I don't know the number of cases in advance. But if you think an ArrayList is better I can change that. Btw `join` is also used do display the guards, where the problem of LinkedList vs ArrayList is similar.